File: web/src/shared/http/envelope.ts
/**
 * The only module that knows the wire envelope. Everything above it sees
 * domain values and `Page<T>`.
 *
 * Contract (docs/contracts/README.md): every 2xx body is `{"data": ...}`;
 * cursor lists add `"meta": {cursor, has_more, limit}`. Cursors are opaque —
 * they are carried, never parsed.
 */

import { MalformedResponseError } from './errors'

export interface Page<T> {
  items: T[]
  /** Opaque. Pass back as the `cursor` query param; never inspect or build one. */
  cursor: string | null
  hasMore: boolean
  limit: number
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

export function unwrapData<T>(body: unknown, parse: (raw: unknown) => T): T {
  if (!isRecord(body) || !('data' in body)) {
    throw new MalformedResponseError('response body is not a {"data": ...} envelope')
  }
  return parse(body.data)
}

export function unwrapPage<T>(body: unknown, parseItem: (raw: unknown) => T): Page<T> {
  if (!isRecord(body) || !('data' in body)) {
    throw new MalformedResponseError('response body is not a {"data": ...} envelope')
  }
  if (!Array.isArray(body.data)) {
    throw new MalformedResponseError('paginated response data is not an array')
  }
  const meta = body.meta
  if (!isRecord(meta) || !('has_more' in meta) || !('limit' in meta)) {
    throw new MalformedResponseError(
      'paginated response is missing meta {cursor, has_more, limit} — ' +
        'a pre-envelope shape such as next_cursor is a contract violation',
    )
  }
  const cursor = meta.cursor
  return {
    items: body.data.map(parseItem),
    cursor: typeof cursor === 'string' ? cursor : null,
    hasMore: meta.has_more === true,
    limit: typeof meta.limit === 'number' ? meta.limit : body.data.length,
  }
}

File: mobile/lib/core/network/envelope.dart
import '../../domain/shared/page.dart';
import 'api_exception.dart';

typedef Parse<T> = T Function(Object? json);

/// The ONLY place the canonical `{"data": ...}` envelope is opened (ADR-010).
/// No widget, notifier or repository ever sees the wrapper.
T unwrapData<T>(Object? body, Parse<T> parse) {
  if (body is! Map<String, Object?> || !body.containsKey('data')) {
    throw MalformedResponseException(body);
  }
  return parse(body['data']);
}

/// Cursor lists add `meta: {cursor, has_more, limit}`. Endpoints that return a
/// plain canonical array with no `meta` — `GET /me/clans` — are treated as a
/// single complete page.
Page<T> unwrapPage<T>(Object? body, Parse<T> parse) {
  if (body is! Map<String, Object?> || body['data'] is! List) {
    throw MalformedResponseException(body);
  }
  final items = (body['data']! as List<Object?>).map(parse).toList();
  final meta = body['meta'];
  if (meta is! Map<String, Object?>) {
    return Page<T>(
      items: items,
      cursor: null,
      hasMore: false,
      limit: items.length,
    );
  }
  return Page<T>(
    items: items,
    // Opaque: stored and replayed verbatim, never parsed.
    cursor: meta['cursor'] as String?,
    hasMore: meta['has_more'] as bool? ?? false,
    limit: meta['limit'] as int? ?? items.length,
  );
}
